control-plane/p4RuntimeSerializer.cpp

```diff
@@ -1636,6 +1641,19 @@ class P4RuntimeEntriesConverter {
         protoRange->set_high(*endStr);
     }
 
+    void addOptional(p4v1::TableEntry* protoEntry, int fieldId,
+                     const IR::Expression* k, int keyWidth,
+                     TypeMap* typeMap) const {
+        if (k->is<IR::DefaultExpression>())  // don't care, skip in P4Runtime message
+            return;
+        auto protoMatch = protoEntry->add_match();
+        protoMatch->set_field_id(fieldId);
+        auto protoOptional = protoMatch->mutable_optional();
+        auto value = convertSimpleKeyExpression(k, keyWidth, typeMap);
+        if (value == boost::none) return;
```

Comment:
I'm not super familiar with this code, but I think this check needs to happen before creating the match (`protoEntry->add_match()`). If value is none, then the match is a wildcard, and should not appear in the P4RT message.

Reply:
I confess that this part of the code I patterned after the `addExact` method elsewhere in the same file, with the primary difference being that `addOptional` checks for the use of a DefaultExpression, but addExact does not (and should not).

@antoninbas Is it a bug if `addExact` is doing this kind of a check and `return`?

Reply:
The code as it is now is correct.

The check at the beginning of the function skips wildcard matches:
```
if (k->is<IR::DefaultExpression>())
    return;
```

This check:
```
if (value == boost::none) return;
```
just skips the field if there is an error in `convertSimpleKeyExpression`.

Reply:
And most likely, if the code reaches this point after many other sanity checks have been done, there might not be any cases left where `convertSimpleKeyExpression` has an error?  At least, I would guess it would be difficult to find a scenario where it does.

Reply:
actually your very own example hits an error in `convertSimpleKeyExpression`: https://github.com/p4lang/p4c/pull/2146/files#diff-a414f2e1b347d4c3f183729cb058e3c1R7

this leads to an empty `optional` match in the P4Runtime entries file: https://github.com/p4lang/p4c/pull/2146/files#diff-f4f53336a8d660ad338f787608f89c6cR8

IMO this is the right thing to do here: we generate an invalid file in case of error rather than treating it the same as a wildcard.

But I think Stefan was just confused as to which check was responsible for handling wildcards.

Reply:
Thanks for all the comments, I was indeed just confused (which is why I said I don't know the code very well) :)